File: leetcode/minimum-path-sum/minimum-path-sum.cpp
/**************************************
* @author dujiong
* @date 2016.11.23
* @version V0.1
**************************************/

class Solution {
public:
    int minPathSum(vector<vector<int> > &grid) {
        int m = grid.size();			//m: 行
        int n = grid[0].size();			//n: 列
        int dp[m][n];
        dp[0][0]=grid[0][0];
        for(int i=1;i<m;i++){
            dp[i][0]=dp[i-1][0]+grid[i][0];
        }
        for(int j=1;j<n;j++){
            dp[0][j]=dp[0][j-1]+grid[0][j];
        }
        for(int i=1;i<m;i++){
            for(int j=1;j<n;j++){
                dp[i][j]=std::min(dp[i][j-1], dp[i-1][j])+grid[i][j];
            }
        }
        return dp[m-1][n-1];
    }
};

File: leetcode-golang/Minimum Path Sum/Minimum Path Sum.go
func minPathSum(grid [][]int) int {
	row := len(grid)
	if row == 0 {
		return 0
	}
	column := len(grid[0])
	for i := 0; i < row; i++ {
		if i >= 1 {
			grid[i][0] += grid[i-1][0]
		}
	}
	for j := 0; j < column; j++ {
		if j >= 1 {
			grid[0][j] += grid[0][j-1]
		}
	}
	for i := 1; i < row; i++ {
		for j := 1; j < column; j++ {
			grid[i][j] = min(grid[i-1][j], grid[i][j-1]) + grid[i][j]
		}
	}
	return grid[row-1][column-1]
}

func min(i, j int) int {
	if i < j {
		return i
	}
	return j
}